libexec/rtld-elf/rtld_c18n.c

```diff
@@ -1436,6 +1453,12 @@ c18n_return_address(void)
 {
 	struct trusted_frame *tframe;
 
+	if (!C18N_ENABLED)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wframe-address"
+		return (__builtin_return_address(1));
```

Comment:
This is indeed a bit fragile. I can wrap this function around a macro which checks `C18N_ENABLED` and returns `__builtin_return_address(0)` straight away.

Reply:
Maybe I should just turn the whole function into a macro? The `__builtin_frame_address(2)` would then become `__builtin_frame_address(1)`, which cannot possibly be mis-optimised because the caller of `dlsym` etc. is a trampoline.